### PR TITLE
TestNode tidyups

### PR DESCRIPTION
**Pull request by jnewbery**
Some additional tidyups after the introduction of TestNode:

- commit 1 makes TestNode use the correct rpc timeout. This should have been included in #11077
- commit 2 separates `add_node()` from `start_node()` as originally discussed here: https://github.com/bitcoin/bitcoin/pull/10556#discussion_r121161453 with @kallewoof . The test writer no longer needs to assign to `self.nodes` when starting/stopping nodes.
- commit 3 adds a `set_test_params()` method, so individual tests don't need to override `__init__()` and call `super().__init__()`

### CODE REVIEW

File: test/functional/txn_clone.py [hunk 1]
```diff
@@ -8,12 +8,6 @@
 from test_framework.util import *
 
 class TxnMallTest(BitcoinTestFramework):
-
-    def __init__(self):
-        super().__init__()
-        self.num_nodes = 4
-        self.setup_clean_chain = False
```
Comment (MarcoFalke) [inline test/functional/txn_clone.py]: I prefer to explicitly set the network topology and initial state of the chain.

File: test/functional/test_framework/test_framework.py [hunk 2]
```diff
@@ -202,51 +175,91 @@ def main(self):
             logging.shutdown()
             sys.exit(TEST_EXIT_FAILED)
 
-    # Public helper methods. These can be accessed by the subclass test scripts.
+    # Methods to override in subclass test scripts.
+    def set_test_params(self):
+        """Override this method to change default values for number of nodes, topology, etc"""
+        pass
```
Comment (MarcoFalke) [inline test/functional/test_framework/test_framework.py]: should raise `RuntimeError('Not implemented')` to enforce the docstring.

**Review by maflcko — COMMENTED** [review]
Concept ACK after quick glance

File: test/functional/test_framework/test_framework.py [hunk 3]
```diff
@@ -202,51 +175,91 @@ def main(self):
...
-        return node
+    def run_test(self):
+        """Override this method to define test logic"""
+        raise NotImplementedError
 
-    def start_nodes(self, num_nodes, dirname, extra_args=None, rpchost=None, timewait=None, binary=None):
-        """Start multiple bitcoinds, return RPC connections to them"""
+    # Public helper methods. These can be accessed by the subclass test scripts.
+
+    def add_nodes(self, num_nodes, dirname, extra_args=None, rpchost=None, timewait=None, binary=None):
```
Comment (kallewoof) [inline test/functional/test_framework/test_framework.py]: Since `self` knows about `self.options.tmpdir`, and all calls to `self.add_nodes` passes `self.options.tmpdir` as `dirname` argument, it seems clean to just give it a default `None` and to set to `self.options.tmpdir if None` in `add_nodes`.
Or perhaps remove it entirely and always use `self.options.tmpdir`. Will it ever be anything else?

File: test/functional/test_framework/test_node.py [hunk 4]
```diff
@@ -61,9 +61,13 @@ def __getattr__(self, *args, **kwargs):
         assert self.rpc_connected and self.rpc is not None, "Error: no RPC connection"
         return self.rpc.__getattr__(*args, **kwargs)
 
-    def start(self):
+    def start(self, extra_args, stderr=None):
```
Comment (kallewoof) [inline test/functional/test_framework/test_node.py]: `extra_args=None, stderr=None` maybe?

File: test/functional/txn_clone.py (hunk 1 above)
Comment (kallewoof) [inline test/functional/txn_clone.py]: Faint agreement. It's easy to forget and start up 4 nodes when you only really need 2, for example. Not seeing the number of nodes makes that even harder to remember.

File: test/functional/wallet.py [hunk 5]
```diff
@@ -7,28 +7,27 @@
...
-    def __init__(self):
-        super().__init__()
+    def set_test_params(self):
         self.setup_clean_chain = True
-        self.num_nodes = 4
         self.extra_args = [['-usehd={:d}'.format(i%2==0)] for i in range(4)]
 
     def setup_network(self):
-        self.nodes = self.start_nodes(3, self.options.tmpdir, self.extra_args[:3])
+        self.add_nodes(4, self.options.tmpdir, self.extra_args)
```
Comment (kallewoof) [inline test/functional/wallet.py]: Why does it not just `add_nodes(3, ..)` here and then add/start/connect the fourth node below? That would remove a lot of necessary `[self.nodes[0:3]]` stuff.

**Review by kallewoof — COMMENTED** [review]
utACK 06bde74f5ec8660975275f8594aee3a99064e327

File: test/functional/txn_clone.py (hunk 1 above)
Comment (jnewbery) [inline test/functional/txn_clone.py]: Sounds good. I've added a new commit that makes setting `self.num_nodes` mandatory. I haven't done the same for `self.setup_clean_chain`, since I think it makes more sense for that to have a default. If people feel strongly, it can be added in a future commit.

File: test/functional/test_framework/test_framework.py (hunk 2 above)
Comment (jnewbery) [inline test/functional/test_framework/test_framework.py]: Yes - done

File: test/functional/test_framework/test_framework.py (hunk 3 above)
Comment (jnewbery) [inline test/functional/test_framework/test_framework.py]: Good point. Thanks! Added this in a new commit.

File: test/functional/test_framework/test_node.py (hunk 4 above)
Comment (jnewbery) [inline test/functional/test_framework/test_node.py]: `TestNode.start` is only ever called by `TestFramework.start_node`, which has its own default for `extra_args`, so not strictly necessary, but I agree having a default here makes it clearer. Added.

File: test/functional/wallet.py (hunk 5 above)
Comment (jnewbery) [inline test/functional/wallet.py]: Because there's not currently a method to add a single node. The model after this PR is to add all nodes at the top of the test and then start/stop them as required during the test.

We could potentially add an `add_node()` method to be called during the test, but lets leave that for a future PR.

**Comment by jnewbery** [comment]
@MarcoFalke @kallewoof - thanks for the review. I think I've responded to all your comments.

New commits pushed, and rebased on master.

**Comment by kallewoof** [comment]
utACK df824a9

File: test/functional/assumevalid.py [hunk 6]
```diff
@@ -54,16 +54,16 @@ def send_header_for_blocks(self, new_blocks):
         self.send_message(headers_message)
 
 class AssumeValidTest(BitcoinTestFramework):
-    def __init__(self):
-        super().__init__()
+    def set_test_params(self):
         self.setup_clean_chain = True
```
Comment (promag) [inline test/functional/assumevalid.py]: It seems there are more `setup_clean_chain = True` so why not make it the default?

**Review by promag — COMMENTED** [review]
Have you thought of not having any node started automatically? There are some tests that start with `stop_node(0)`..

So each`run_test(self)` the nodes are explicitly started, `num_nodes` can go away, even `setup_clean_chain` would be a `start_node` argument.

File: test/functional/assumevalid.py (hunk 6 above)
Comment (jnewbery) [reply test/functional/assumevalid.py]: I haven't made any changes to `setup_clean_chain`. See earlier comment:

> I haven't done the same for self.setup_clean_chain, since I think it makes more sense for that to have a default. If people feel strongly, it can be added in a future commit.

Feel free to open a PR to change/remove the default if you think that's necessary. I don't think it's required for this PR.

**Comment by jnewbery** [comment]
Thanks for the reivew @promag 

> Have you thought of not having any node started automatically? There are some tests that start with stop_node(0)

Tests that don't need their nodes to start should override the `setup_network()` or `setup_nodes()` methods. The vast majority of tests require nodes to be started at the top of the test, so that's the default.

> setup_clean_chain would be a start_node argument.

`setup_clean_chain` is a property of the pre-mined chain. I don't think we should change it to be a property of the individual node (if you disagree please open a separate PR to demonstrate the change).

File: test/functional/bumpfee.py [hunk 7]
```diff
@@ -34,21 +34,18 @@ def __init__(self):
...
-        self.nodes = self.start_nodes(self.num_nodes, self.options.tmpdir, extra_args)
-
+    def run_test(self):
         # Encrypt wallet for test_locked_wallet_fails test
         self.nodes[1].node_encrypt_wallet(WALLET_PASSPHRASE)
-        self.nodes[1] = self.start_node(1, self.options.tmpdir, extra_args[1])
+        self.start_node(1)
         self.nodes[1].walletpassphrase(WALLET_PASSPHRASE, WALLET_PASSPHRASE_TIMEOUT)
 
         connect_nodes_bi(self.nodes, 0, 1)
```
Comment (MarcoFalke) [inline test/functional/bumpfee.py]: This is already done by `setup_network`

Comment (jnewbery) [reply test/functional/bumpfee.py]: L41 stops node 1 (the `encryptwallet` RPC shuts down the node), so I believe all the p2p connections are lost and we need to reconnect the nodes after restarting node 1.

I've tested removing these two lines and the test fails (the next call to `sync_all()` fails)

Comment (MarcoFalke) [reply test/functional/bumpfee.py]: Sorry, missed that the node shuts down.

File: test/functional/pruning.py [hunk 8]
```diff
@@ -307,15 +313,15 @@ def has_block(index):
 
         # stop node, start back up with auto-prune at 550MB, make sure still runs
         self.stop_node(node_number)
-        self.nodes[node_number] = self.start_node(node_number, self.options.tmpdir, ["-prune=550"], timewait=900)
+        self.start_node(node_number, extra_args=["-prune=550"])
```
Comment (MarcoFalke) [inline test/functional/pruning.py]: You are removing `timewait=900` from all calls but the first. I assume this does not cause issues?

File: test/functional/wallet.py [hunk 9]
```diff
@@ -208,8 +211,10 @@ def run_test(self):
         zeroValueTxid= decRawTx['txid']
         sendResp = self.nodes[1].sendrawtransaction(signedRawTx['hex'])
 
+        # self.sync_all([self.nodes[0:3]])
```
Comment (MarcoFalke) [inline test/functional/wallet.py]: debug statement?

File: test/functional/wallet.py [hunk 10]
```diff
@@ -208,8 +211,10 @@ def run_test(self):
...
         self.sync_all()
         self.nodes[1].generate(1) #mine a block
+        # self.sync_all([self.nodes[0:3]])
```
Comment (MarcoFalke) [inline test/functional/wallet.py]: debug statement?

File: test/functional/wallet.py [hunk 11]
```diff
@@ -247,14 +254,16 @@ def run_test(self):
...
         self.stop_nodes()
-        self.nodes = self.start_nodes(3, self.options.tmpdir)
+        self.start_node(0)
+        self.start_node(1)
+        self.start_node(2)
         connect_nodes_bi(self.nodes,0,1)
         connect_nodes_bi(self.nodes,1,2)
         connect_nodes_bi(self.nodes,0,2)
-        sync_blocks(self.nodes)
+        sync_blocks(self.nodes[0:2])
```
Comment (MarcoFalke) [inline test/functional/wallet.py]: why not `self.nodes[:3]`?

File: test/functional/wallet.py [hunk 12]
```diff
@@ -247,14 +254,16 @@ def run_test(self):
...
 
         self.nodes[0].generate(1)
-        sync_blocks(self.nodes)
+        sync_blocks(self.nodes[0:2])
```
Comment (MarcoFalke) [inline test/functional/wallet.py]: why not `self.nodes[:3]`?

File: test/functional/zapwallettxes.py [hunk 13]
```diff
@@ -20,11 +20,9 @@
                                  )
 
 class ZapWalletTXesTest (BitcoinTestFramework):
-
-    def __init__(self):
-        super().__init__()
+    def set_test_params(self):
         self.setup_clean_chain = True
-        self.num_nodes = 2
+        self.num_nodes = 3
```
Comment (MarcoFalke) [inline test/functional/zapwallettxes.py]: why is this changed in a refactoring commit?

**Review by maflcko — COMMENTED** [review]
utACK df824a92121eb68849e474e87098a441d4e52926

File: test/functional/pruning.py (hunk 8 above)
Comment (jnewbery) [reply test/functional/pruning.py]: `rpc_timeout` is a property of `TestNode` set when instantiating the python object, not when starting up the bitcoind node.

File: test/functional/wallet.py (hunk 9 above)
Comment (jnewbery) [reply test/functional/wallet.py]: derp. Fixed

File: test/functional/wallet.py (hunk 10 above)
Comment (jnewbery) [reply test/functional/wallet.py]: yep

File: test/functional/wallet.py (hunk 11 above)
Comment (jnewbery) [reply test/functional/wallet.py]: fixed

File: test/functional/wallet.py (hunk 12 above)
Comment (jnewbery) [reply test/functional/wallet.py]: fixed

File: test/functional/zapwallettxes.py (hunk 13 above)
Comment (jnewbery) [reply test/functional/zapwallettxes.py]: fixed

**Comment by jnewbery** [comment]
Thanks @MarcoFalke . Review comments addressed.

**Comment by MarcoFalke** [comment]
re-utACK 7148b74dc39110f53c665b94fa9d994c6ad6dc1c
